Sources/AWSSDKSwiftCore/Credential/RuntimeCredentialProvider.swift

```diff
@@ -0,0 +1,122 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the AWSSDKSwift open source project
+//
+// Copyright (c) 2017-2020 the AWSSDKSwift project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of AWSSDKSwift project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOConcurrencyHelpers
+
+class RuntimeCredentialProvider: CredentialProvider {
+    
+    static func createProvider(
+        on eventLoop: EventLoop,
+        httpClient: AWSHTTPClient) -> CredentialProvider
+    {
+        // 1. has the environment credentials? let's use those...
+        //    we are creating static environment credentials here to save the lock in runtime
+        //    credential scenarios.
+        if let credential = StaticCredential.fromEnvironment() {
+            return credential
+        }
+        
+        // 2. we don't have static credentials. let's determine what to do while running
+        return RuntimeCredentialProvider(eventLoop: eventLoop, httpClient: httpClient)
+    }
+    
+    let lock = Lock()
+    var internalProvider: CredentialProvider? {
+        get {
```

Comment:
Any reason for not adding a setter here using the lock instead of calling withLock everywhere you set the `_internal` value. You could even break this out, into a mini struct that holds the lock, or if you're being clever implement a property wrapper that provides locking access.

Reply:
I want the `internalProvider` to be a private set.

> instead of calling withLock everywhere

Well, it's one occurrence so far.

Reply:
You could use `private(set)`